### PR TITLE
fix(kds): don't delete zone created Secret

### PR DESCRIPTION
**Pull request by lukidzi**
## Supporting documentation

Fix #XX

### CODE REVIEW

File: pkg/kds/v2/client/kds_client.go [hunk 1]
```diff
@@ -112,7 +113,19 @@ func (s *kdsSyncClient) Receive() error {
...
+				if err := s.kdsStream.NACK(received.Type, err); err != nil {
+					if err == io.EOF {
+						return nil
+					}
+					return errors.Wrap(err, "failed to NACK a discovery response")
+				}
+				s.log.V(1).Info("NACK", "response", received)
+				continue
+			} else {
+				return errors.Wrapf(err, "failed to store %s resources", received.Type)
```
Comment (lahabana) [inline pkg/kds/v2/client/kds_client.go:127]: should we include the resource name too?

**Comment by lahabana** [comment]
2 points:

- needs test
- Can we consider backporting this?

File: api/system/v1alpha1/secret.proto [hunk 2]
```diff
@@ -15,6 +15,7 @@ message Secret {
   option (kuma.mesh.resource).package = "system";
   option (kuma.mesh.resource).skip_kubernetes_wrappers = true;
   option (kuma.mesh.resource).kds.send_to_zone = true;
+  option (kuma.mesh.resource).kds.send_to_global = false;
```
Comment (Icarus9913) [inline api/system/v1alpha1/secret.proto:18]: Does this supplement make impact? I don't see any place call it.
Or is it just a standard supplement to the structure?

File: pkg/kds/v2/client/kds_client.go [hunk 3]
```diff
@@ -112,7 +113,19 @@ func (s *kdsSyncClient) Receive() error {
...
+			}
```
Comment (Icarus9913) [inline pkg/kds/v2/client/kds_client.go:128]: Should we filter this `ResourceAlreadyExists` error when creating and then downgrade the error? I feel confused if we send NACK, will we do this create again in the next streaming?

Comment (lukidzi) [reply pkg/kds/v2/client/kds_client.go]: I've changed the logic

File: api/system/v1alpha1/secret.proto (hunk 2 above)
Comment (lukidzi) [reply api/system/v1alpha1/secret.proto:18]: It makes no impact since the default is `false`. I wanted to make it explicit

**Comment by lukidzi** [comment]
> * needs test
Added both E2E and unit
> * Can we consider backporting this?
I think it's possible to backport this and https://github.com/kumahq/kuma/pull/12776. I am a bit worried since it changes logic a bit.

**Comment by Icarus9913** [comment]
> Do we need kuma.io/origin validation?

In my opinion: no

The cases we may hit:
1. create secret in zone (restart ZCP)
2. create secret in global and it would sync to zone (restart GCP, ZCP)
3. create secret in zone and create the same name secret in global (restart ZCP, GCP)

Since we have used `predicate` to filter the zone secret, it makes no impact for the `kuma.io/origin: global` global synced secret right? Do we have other edge cases?

File: pkg/kds/v2/store/sync.go [hunk 4]
```diff
@@ -41,12 +42,13 @@ type ResourceSyncer interface {
 	//
 	// Sync takes into account only 'Name' and 'Mesh' when it comes to upstream's Meta.
 	// 'Version', 'CreationTime' and 'ModificationTime' are managed by downstream store.
-	Sync(ctx context.Context, upstream client_v2.UpstreamResponse, fs ...SyncOptionFunc) error
+	Sync(ctx context.Context, upstream client_v2.UpstreamResponse, fs ...SyncOptionFunc) (error, error)
```
Comment (Icarus9913) [inline pkg/kds/v2/store/sync.go:45]: Can we add comments for these 2 errors since this is a public method

**Comment by lukidzi** [comment]
> The cases we may hit:
>
>1. create secret in zone (restart ZCP)
I didn't add this specific e2e tests since creating a new secret triggers reconciliation + restarting CP might be flaky in the e2e
>2. create secret in global and it would sync to zone (restart GCP, ZCP)
We should manually test
>3. create secret in zone and create the same name secret in global (restart ZCP, GCP)
We should manually test

>Since we have used predicate to filter the zone secret, it makes no impact for the kuma.io/origin: global global synced >secret right? Do we have other edge cases?

I think it should all works and don't affect.

File: test/e2e_env/multizone/sync/sync.go [hunk 5]
```diff
@@ -187,6 +213,84 @@ spec:
...
+
+			Expect(multizone.Global.Install(YamlUniversal(globalSecret))).To(Succeed())
+			// should not override existing in zone
+			Eventually(func(g Gomega) {
+				out, err := multizone.Global.GetKumactlOptions().RunKumactlAndGetOutput("get", "secrets", "--mesh", meshName, "-o", "yaml")
+				g.Expect(err).ToNot(HaveOccurred())
+				g.Expect(strings.Count(out, secretName)).To(Equal(1))
+				g.Expect(strings.Count(out, "Z2xvYmFsCg==")).To(Equal(1))
+			}, "30s", "1s").Should(Succeed())
+			Consistently(func(g Gomega) {
```
Comment (Icarus9913) [inline test/e2e_env/multizone/sync/sync.go:260]: Do we need to use `Consistently` here for Zone? That would lead to a flaky error once network latency. Maybe `Eventually` is enough right?

**Review by Icarus9913 — APPROVED** [review]
